HOTT-3512: Add service-level Terraform

amberstarlight opened the pull request:
### Jira link

[HOTT-3512](https://transformuk.atlassian.net/browse/HOTT-3512)

### What?

I have added/removed/altered:

- Added the service-level Terraform to this repository, under `terraform/`.
- Added a pre-commit file under `terraform/` which is scoped to this subdirectory only.

### Why?

I am doing this because:

- We're now at the point where we can add the Terraform itself to control the service deployment and sub-resources.

> **Note**
> This PR **_DOES NOT_** add pipeline changes to deploy this into any environments, it is just the Terraform itself.

### Extra Info

When this is merged, I will add `[no ci]` or equivalent so that it does not trigger any pipeline, as there's no changes to make.

dependabot[bot] committed:
HOTT-3512: Create application level Terraform

amberstarlight committed:
Merge branch 'main' into hott-3512-add-terraform

amberstarlight committed:
HOTT-3512: Update service module

amberstarlight committed:
HOTT-3512: Make TODOs more explicit

amberstarlight committed:
HOTT-3512: Remove VCAP... env var

This is a CloudFoundry specific environment variable. There's handling
code in lib/paas_config.rb to fall back to REDIS_URL for this purpose.

I've also moved the pre-commit config up one directory. It's scoped to
run hooks only on terraform/* for now so won't create huge diffs, but
it has to be in the root directory to run on the cli.

willfish commented on terraform/locals.tf:
```diff
@@ -0,0 +1,9 @@
+    xi = "https://${var.base_domain}/xi"
+  }
+
+  # TODO: Revisit this when PaaS is no more
```
So this is actually a hard requirement coming from the plek gem which is a nested dependency of one of our other govuk gems https://github.com/alphagov/plek

willfish replied:
So we'll never get rid of it :(

amberstarlight replied:
Makes sense. I'll remove the TODO 😢

willfish commented on terraform/main.tf:
```diff
@@ -0,0 +1,136 @@
+    # TODO: discuss this internally via HOTT-3555
+    {
+      name  = "API_SERVICE_BACKEND_OPTIONS"
+      value = jsonencode(local.api_service_backend_url_options)
```
Yeah, I guess this will get discovered later.

amberstarlight committed:
[no ci] HOTT-3512: Remove TODO from locals

This env var is actually a hard requirement coming from the plek gem
which is a nested dependency of one of our other govuk gems
https://github.com/alphagov/plek

willfish reviewed (approved):
Excited to get this rolling out :tada:
